### DISCUSSION TITLE
Diagram Plugin

### DISCUSSION THREAD

**Discussion by itdependsnetworks** [nautobot/nautobot#1012] [Plugin (App) Ideas & Development]
### User Story

As **Nelly the Network Engineer**
**I want** to view my network infrastructure visually, as a diagram
**So that** I can get rid of my manually created visio diagrams
**I will know this is done** when I have a visual diagram that represents my network 

### Database Changes (Optional)

No

### External Dependencies (Optional)

Likely need to include the external library as the diagram tool.

### Other

* This is likely handled best by a plugin
* Need to consider if diagrams can be only manually generated, dynamically generated, or both
* Need to consider the output of the diagrams, is it compatible with other services? 
* Libraries to consider
  * https://n2g.readthedocs.io/en/latest/
  * https://diagrams.mingrammer.com/
  * https://d3js.org/

**Comment by glennmatthews** [comment]
Worth considering/investigating whether `nextbox-ui-plugin` (https://github.com/iDebugAll/nextbox-ui-plugin) could be ported to support Nautobot, and whether that meets this need.

**Comment by itdependsnetworks** [comment]
I do not believe the NeXt UI is supported anymore. 
![image](https://user-images.githubusercontent.com/9260483/136097128-35423197-aec0-49ad-aba4-2bf30a468ec5.png)

**Comment by jvanderaa** [comment]
Perhaps this should be moved to a discussion for further.

I have been thinking about this less formally and I would think that the diagrams package would work well with the Affinity icon set: https://github.com/ecceman/affinity

**Comment by jedelman8** [comment]
We created a Discussion Category around Plugin Development and will be moving this there!

**Comment by N0odlez** [comment]
Hello everyone,

I was wondering if there may be any traction on this becoming a feature/plugin?

We all hate documentation and making Visio diagrams is the bane of our lives. Wouldn't it be great to export physical and logical diagrams from Nautobot.

**Comment by jpmorigin** [comment]
Would it be possible to port this over: https://github.com/mattieserver/netbox-topology-views

**Reply by N0odlez** [reply]
This would be ideal!

**Comment by hvisage** [comment]
Busy with evaluating and this is something that came up, so is there any traction/work even if alpha for diagrams from links states in nautobot?

**Comment by FengLovesBach** [comment]
Hello, does anyone when this feature might be available? Thanks.

**Comment by htam-net** [comment]
Hello,

For sharing : [https://gitlab-ce.gwdg.de/gwdg-netz/nautobot-plugins/nautobot-ui-plugin](https://gitlab-ce.gwdg.de/gwdg-netz/nautobot-plugins/nautobot-ui-plugin) 
based on : [https://github.com/iDebugAll/nextbox-ui-plugin ](https://github.com/iDebugAll/nextbox-ui-plugin )

And listed in the community-developped Apps : [https://docs.nautobot.com/projects/core/en/stable/apps/nautobot-apps/](https://docs.nautobot.com/projects/core/en/stable/apps/nautobot-apps/)

**Reply by gioccher** [reply]
unfortunately since nautobot 2 it has been a bit clunky, and on nautobot 3 it doesn't work anymore `ImportError: cannot import name 'PluginConfig' from 'nautobot.extras.plugins'`

**Comment by gioccher** [comment]
I ported the netbox topology views plugin to nautobot 3 https://github.com/gioccher/nautobot-topology-views
It works well enough for my needs, but I have not really tried all the functionality (your mileage may vary)
<img width="2560" height="1944" alt="topology_light" src="https://github.com/user-attachments/assets/3251cb95-1945-4401-9e9d-62358be4c677" />
